pkg/secureenclave/secureenclave.m

```diff
@@ -0,0 +1,271 @@
+    *ret = CFDataToUint8(publicKeyRef);
+    CFIndex size = CFDataGetLength(publicKeyRef);
+    CFRelease(publicKeyRef);
+    publicKeyRef = NULL;
+    return size;
+}
+
+OSStatus findPrivateKey(CFDataRef pubKeySha1, SecKeyRef *key) {
+    NSDictionary* attributes =
+    @{ (id)kSecClass: (id)kSecClassKey,
+        (id)kSecAttrKeyType: (id)kSecAttrKeyTypeEC,
+        (id)kSecAttrTokenID: (id)kSecAttrTokenIDSecureEnclave,
+        (id)kSecAttrKeyClass: (id)kSecAttrKeyClassPrivate,
+        (id)kSecReturnRef: (id)kCFBooleanTrue,
+        (id)kSecMatchLimit: (id)kSecMatchLimitOne,
+        (id)kSecAttrApplicationLabel: (__bridge id)pubKeySha1,
+    };
+
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)attributes, (__bridge void *)key);
+    CFRelease(attributes);
+    attributes = NULL;
+    return status;
+}
+
+size_t findKey(unsigned char* hash, unsigned char** ret, char** retErr){
+    #define sha1HashSize 20
+    CFDataRef cfHash = CFDataCreateWithBytesNoCopy(kCFAllocatorDefault, (UInt8*)hash, sha1HashSize, kCFAllocatorNull);
+    SecKeyRef privateKey = NULL;
+    OSStatus status = findPrivateKey(cfHash, (__bridge void *)&privateKey);
+
+    CFRelease(cfHash);
+    cfHash = NULL;
+
+    if ((status != 0) || (!privateKey)) {
+      NSString *msg = [NSString stringWithFormat:@"finding key pair: status %i", (int)status];
+      *retErr = CFStringToCString((__bridge CFStringRef)msg);
+      CFRelease(msg);
+      msg = NULL;
+      return 0;
+    }
+
+    // get public key from private key
+    SecKeyRef publicKey = SecKeyCopyPublicKey(privateKey);
+    CFRelease(privateKey);
+    privateKey = NULL;
+
+    if (!publicKey) {
+      CFStringRef errStr = CFSTR("failed to copy public key");
+      *retErr = CFStringToCString(errStr);
+      CFRelease(errStr);
+      errStr = NULL;
+      return 0;
+    }
+
+    // extract just the public key data
+    CFDataRef publicKeyRef = ExtractPubKey(publicKey);
+    *ret = CFDataToUint8(publicKeyRef);
+    CFIndex size = CFDataGetLength(publicKeyRef);
+    CFRelease(publicKeyRef);
+    publicKeyRef = NULL;
+    return size;
+}
+
+size_t sign(unsigned char* hash, unsigned char* data, int dataSize, unsigned char** ret, char** retErr){
+    CFDataRef cfHash = CFDataCreateWithBytesNoCopy(kCFAllocatorDefault, (UInt8*)hash, 20, kCFAllocatorNull);
+    SecKeyRef privateKey = NULL;
+    OSStatus status = findPrivateKey(cfHash, (__bridge void *)&privateKey);
+
+    if ((status != 0) || (!privateKey)) {
+      NSString *msg = [NSString stringWithFormat:@"finding key: status %i", (int)status];
+      *retErr = CFStringToCString((__bridge CFStringRef)msg);
+      CFRelease(msg);
+      msg = NULL;
+      return 0;
+    }
+
+    CFDataRef dataRef = CFDataCreateWithBytesNoCopy(kCFAllocatorDefault, (UInt8*)data, dataSize, kCFAllocatorNull);
+    CFErrorRef error = NULL;
+
+    CFDataRef signature = SecKeyCreateSignature(privateKey, kSecKeyAlgorithmECDSASignatureDigestX962, dataRef, &error);
```

Comment:
`SecKeyCreateSignature` docs don't specify what happens, but we may want to avoid calling if `dataRef` is NULL... or it might just be handled naturally by the error handling below?

Reply:
no null checking